frontend/src/metabase/selectors/metadata.js

```diff
@@ -104,16 +104,9 @@ export const getMetadata = createSelector(
     );
     // hydrateList(meta.schemas, "tables", meta.tables);
     hydrate(meta.schemas, "tables", schema =>
-      schema.tables
-        ? // use the schema tables if they exist
-          schema.tables.map(t => meta.table(t))
-        : schema.database && schema.database.tables.length > 0
-        ? // if the schema has a database with tables, use those
-          schema.database.tables.filter(t => t.schema_name === schema.name)
-        : // otherwise use any loaded tables that match the schema id
-          Object.values(meta.tables).filter(
-            t => t.schema && t.schema.id === schema.id,
-          ),
+      Object.values(meta.tables).filter(
+        t => t.schema && t.schema.id === schema.id,
+      ),
```

Comment:
@tlrobinson I wanted to get your thoughts on dropping those optimizations.

With this PR's change, hidden tables are only loaded in the datamodel pages. If a user has loaded tables in the main app and then goes to the datamodel we wouldn't show the full list. It seems that this bug occurred because even though the full table list was loaded, the `schema.tables` and `database.tables` lists weren't updated. Removing this optimization fixed it.

Do you think removing this will kill performance?

Reply:
Aside from main problem I described already, I do think in general it's better to have a single "source of truth", so this might be good to do. A possible downside is it loses the ordering provided by the API, if that matters.

I could see this being slow if there are lots of schemas/tables. One optimization to avoid iterating the whole list of tables for every schema would be to pre-group tables by schema, something like:

```js
const tablesBySchema = _.groupBy(meta.tables, (id, table) => table.schema && table.schema.id);
hydrate(meta.schemas, "tables", schema => tablesBySchema[schema.id] || [])
```

We should also make sure table objects returned by the server always have a `schema` property (or they're added before normalizing like in .